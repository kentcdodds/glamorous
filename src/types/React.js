// @flow
import type {Component, Element} from 'react'

export type FunctionalComponent = (
  {[key: string]: any}, // props
  ?{[key: string]: any}, // context
) => Element<*> | null;

// eslint-disable-next-line no-undef
export type ComponentClass = Class<Component<*, *, *>>;

export type RefCallback<Instance> = (Instance | null) => void;
