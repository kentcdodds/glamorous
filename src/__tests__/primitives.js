import React from 'react'
import {render, mount} from 'enzyme'
import * as jestGlamorReact from 'jest-glamor-react'
import {StyleSheet} from 'react-primitives'
import glamorous from '../primitives/index'

expect.extend(jestGlamorReact.matcher)
expect.addSnapshotSerializer(jestGlamorReact.serializer)

describe('Primitive interfaces', () => {
  test('sanity test', () => {
    const Component = glamorous.view({margin: '10px'})
    expect(<Component />).toMatchSnapshotWithGlamor()
  })

  test('should not throw any error when called', () => {
    glamorous.view({})
  })

  test('should combine with StyleSheet.create() (inline styles)', () => {
    const Component = glamorous.view({
      margin: '10px',
      justifyContent: 'center',
      alignItems: 'center',
    })

    const styles = StyleSheet.create({
      foo: {
        backgroundColor: 'red',
        display: 'flex',
      },
    })

    expect(<Component style={styles.foo} />).toMatchSnapshotWithGlamor()
  })

  test('should attach a displayName', () => {
    const Component = glamorous.view({})
    expect(Component.displayName).toBe('glamorous(View)')
  })

  test('should pass the ref to the component', () => {
    const Component = glamorous.view({})
    const ref = jest.fn()

    const wrapper = mount(<Component innerRef={ref} />)
    const View = wrapper.find('View').first()

    expect(ref).toHaveBeenCalledWith(View.node)
  })

  test('should render a component with theme properties', () => {
    const Component = glamorous.text(
      {
        color: 'red',
      },
      (props, theme) => ({padding: theme.padding}),
    )
    expect(
      render(<Component theme={{padding: '10px'}} />),
    ).toMatchSnapshotWithGlamor()
  })

  test('should pass an updated theme when theme prop changes', () => {
    const Component = glamorous.text(
      {
        color: 'red',
      },
      (props, theme) => ({padding: theme.padding}),
    )
    const wrapper = mount(<Component theme={{padding: 10}} />)
    expect(wrapper).toMatchSnapshotWithGlamor(
      `with theme prop of padding 10px`,
    )
    wrapper.setProps({theme: {padding: 20}})
    expect(wrapper).toMatchSnapshotWithGlamor(
      `with theme prop of padding 20px`,
    )
  })

  test('should render the Native <Image /> component', () => {
    const ReactImage = glamorous.image({padding: '10px'})
    expect(
      <ReactImage source="https://facebook.github.io/react/img/logo_og.png" />,
    ).toMatchSnapshotWithGlamor()
  })

  test('combine inline styles with <Image /> component', () => {
    const styles = StyleSheet.create({
      image: {
        justifyContent: 'center',
        alignItems: 'center',
      },
    })
    const ReactImage = glamorous.image({padding: '10px'})
    expect(
      <ReactImage
        style={styles.image}
        source="https://facebook.github.io/react/img/logo_og.png"
      />,
    ).toMatchSnapshotWithGlamor()
  })

  // test('should support jsx-styled api', () => {
  //   const Comp = glamorous(reactPrimitives['View'])()

  //   expect(
  //     render(
  //       <Comp color="red">Hello !!</Comp>
  //     )
  //   ).toMatchSnapshotWithGlamor()
  // })
})
