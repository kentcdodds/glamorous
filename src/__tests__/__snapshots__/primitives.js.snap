// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primitive interfaces combine inline styles with <Image /> component 1`] = `
<glamorous(Image)
  source="https://facebook.github.io/react/img/logo_og.png"
  style={66}
/>
`;

exports[`Primitive interfaces sanity test 1`] = `<glamorous(View) />`;

exports[`Primitive interfaces should combine with StyleSheet.create() (inline styles) 1`] = `
<glamorous(View)
  style={65}
/>
`;

exports[`Primitive interfaces should render a component with theme properties 1`] = `
.css-1fyph4i,
[data-css-1fyph4i] {
  color: red;
  padding: 10px;
}

<div
  class="css-1fyph4i rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-color-homxoj rn-display-1471scf rn-font-1lw9tu2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
  dir="auto"
/>
`;

exports[`Primitive interfaces should render the Native <Image /> component 1`] = `
<glamorous(Image)
  source="https://facebook.github.io/react/img/logo_og.png"
/>
`;

exports[`with theme prop of padding 10px 1`] = `
.css-1p841yy,
[data-css-1p841yy] {
  color: red;
  padding: 10px;
}

<glamorous(Text)
  theme={
    Object {
      "padding": 10,
    }
  }
>
  <Text
    className="css-1p841yy"
  >
    <div
      className="css-1p841yy rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-color-homxoj rn-display-1471scf rn-font-1lw9tu2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
    />
  </Text>
</glamorous(Text)>
`;

exports[`with theme prop of padding 20px 1`] = `
.css-1e41o48,
[data-css-1e41o48] {
  color: red;
  padding: 20px;
}

<glamorous(Text)
  theme={
    Object {
      "padding": 20,
    }
  }
>
  <Text
    className="css-1e41o48"
  >
    <div
      className="css-1e41o48 rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-color-homxoj rn-display-1471scf rn-font-1lw9tu2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
    />
  </Text>
</glamorous(Text)>
`;
